services/s3/src/main/java/software/amazon/awssdk/services/s3/checksums/ChecksumCalculatingAsyncRequestBody.java

```diff
@@ -54,6 +54,9 @@ public void subscribe(Subscriber<? super ByteBuffer> s) {
                                       SdkChecksum sdkChecksum) {
             this.wrapped = wrapped;
             this.checksum = sdkChecksum;
+            // Note: there may be multiple places that hold references to this checksum object, but assumption is that in cse of
+            // retry all 'writes' to this will be done before new attempt to upload is run.
+            checksum.reset();
```

Comment:
It's a bit weird that `SdkChecksum` is injected, if we'll be resetting it here, don't you think? Should we create a new SdkChecksum for each `subscribe` call? I know that involves a bit of refactoring, but it seems cleaner overall.

LMK if you don't want to invest this time and we can take a crack at it when we get a chance.

Reply:
I would agree that this design is weird.
I do not think I'll have to for this at the moment, unfortunately.
However it would be really appreciate the fix being released sooner rather than later - the bug has been opened for quite a while and is quite serious in a sense that it may lead to data being lost.

Reply:
I understand. I'll pick up this change and ferry it through.

Reply:
Yeah, it's definitely too much work to refactor: https://github.com/aws/aws-sdk-java-v2/pull/1552